x-pack/docs/en/security/authorization/custom-authorization.asciidoc

```diff
@@ -0,0 +1,159 @@
+  operations.  It is the responsibility of the implementation to ensure asynchronous behavior
+  and non-blocking calls, which is made easier by the fact that the `ActionListener` is
+  provided on which to send the response when the roles have been resolved and the response
+  is ready.
+
+To package your custom roles provider as a plugin:
+
+. Implement an extension class for your roles provider that implements
+  `org.elasticsearch.xpack.core.security.SecurityExtension`. There you need to
+  override one or more of the following methods:
++
+[source,java]
+----------------------------------------------------
+@Override
+public List<BiConsumer<Set<String>, ActionListener<Set<RoleDescriptor>>>>
+getRolesProviders(Settings settings, ResourceWatcherService resourceWatcherService) {
+    ...
+}
+----------------------------------------------------
++
+The `getRolesProviders` method is used to provide a list of custom roles providers that
+will be used to resolve role names, if the role names could not be resolved by the reserved
+roles or native roles stores.   The list should be returned in the order that the custom role
+providers should be invoked to resolve roles.  For example, if `getRolesProviders` returns two
+instances of roles providers, and both of them are able to resolve role `A`, then the resolved
+role descriptor that will be used for role `A` will be the one resolved by the first roles
+provider in the list.
+
+[[implementing-authorization-engine]]
+==== Implementing an authorization engine
+
+To create an authorization engine, you need to:
+
+. Implement the `org.elasticsearch.xpack.core.security.authz.AuthorizationEngine`
+  interface in a class with the desired authorization behavior.
+. Implement the `org.elasticsearch.xpack.core.security.authz.Authorization.AuthorizationInfo`
+  interface in a class that contains the necessary information to authorize the request.
+
+To package your authorization engine as a plugin:
+
+. Implement an extension class for your authorization engine that extends
+  `org.elasticsearch.xpack.core.security.SecurityExtension`. There you need to
+  override the following method:
++
+[source,java]
+----------------------------------------------------
+@Override
+public AuthorizationEngine getAuthorizationEngine(Settings settings) {
+    ...
+}
+----------------------------------------------------
++
+The `getAuthorizationEngine` method is used to provide the authorization engine
+implementation.
+
+Sample code that illustrates the structure and implementation of a custom
+authorization engine is provided in the
+https://github.com/elastic/elasticsearch/tree/master/plugin/examples/security-example-authorization-engine[elasticsearch]
+repository on GitHub. You can use this code as a starting point for creating your
+own authorization engine.
+
+[[packing-extension-plugin]]
+==== Implement an elasticsearch plugin
+
+In order to register the security extension, you need to also implement an
```

Comment:
I think this section applies to both the previous sections, but maybe this could be stated explicitly.  For example:
```suggestion
In order to register the security extension for your custom roles provider or authorization engine, you need to also implement an
```